README.md
# Pult [![NPM version](https://badge.fury.io/js/pult.png)](http://badge.fury.io/js/pult)

Access local servers on .dev domains.

View the [documentation](http://cmcenroe.me/pult) for more detailed
information.

## How do I get it?

```
npm install pult
```

## How do I use it?

Pult can either be used as a library from within your Node HTTP app, or
from the command line for use with other HTTP servers.

### As a Library

See the [library documentation](http://cmcenroe.me/pult/index.html) for
more information.

```js
var pult = require('pult');
var http = require('http');

var server = http.createServer();

pult.getPort('my-app', function(err, port, domain) {
  if (err) throw err;
  server.listen(port);
  console.log('http://' + domain); // http://my-app.dev
});
```

This example assumes that the Pult server is running. For more detailed
examples and better practices, see the [example
documentation](http://cmcenroe.me/pult/example.html).

### From the Command Line

```
npm install -g pult
```

Running `pult` from the command line automatically spawns `pult-server`
if it is not already running.

```sh
ruby-project $ pult rackup # http://ruby-project.dev
other-project $ pult foreman start # http://other-project.dev
```

By default, Pult sets the `PORT` environment variable then spawns your
server process. If your server process takes a port using the `-p` or
`-P` options rather than through `PORT`, simply pass the required option
to Pult:

```sh
website $ pult -P jekyll serve # http://website.dev
```

For more details about the command line implementation, see the [command
line documentation](http://cmcenroe.me/pult/pult.html).

#### Custom domains

By default, Pult uses the name of the current directory to determine
what `.dev` domain to serve from. You can specify a different domain
using the `-n` option:

```sh
node-project $ pult -n my-project node app # http://my-project.dev
```

#### Subdomains

Pult also serves the same application on all subdomains of its `.dev`
domain. For example, `node-project` will be available at
`http://node-project.dev`, `http://www.node-project.dev`, etc.

Pult also allows you to serve different applications on subdomains
simply by specifying the subdomain with `-n`:

```sh
other-project $ pult -n other.project node app # http://other.project.dev
```

#### Configuration

Pult interprets the contents of `.pultrc` in the current directory as
command-line arguments. This allows saving the domain name and command
to run so that bringing a server up on a `.dev` domain is as simple as
running `pult`.

```sh
node-project $ echo "-n my-project node app" > .pultrc
node-project $ pult # node app on http://my-project.dev
```

#### Status

To find out which domains are being forwarded to which ports, simply run
`pult` with no arguments:

```sh
~ $ pult
[pult] next: 7003
[pult] http://node-project.dev --> http://localhost:7001
[pult] http://ruby-project.dev --> http://localhost:7002
```

#### Output

`pult` logs to standard error prefixed with `[pult]`. To disable
logging, pass the `-q` option.

### Server

By default, `pult-server` respawns itself with sudo then forks to the
background. To run `pult-server` as a foreground process, pass the `-f`
option:

```sh
~ $ pult-server -f
```

To stop `pult-server` while it is running in the background, pass the
`-k` option to `pult`:

```sh
~ $ pult-server
~ $ pult -k
```

The default first port that `pult-server` assigns is 7001. To change
this port, pass the `-p` option:

```sh
~ $ pult-server -p 8081
~ $ pult
[pult] next: 8081
```

By default, `pult-server` binds to ports 53 (DNS) and 80 (HTTP) on
`127.0.0.1` (IPv4 localhost). If you are already running these services
on `127.0.0.1` and are using a system that allows the use of the entire
`127.0.0.0/8` range, you can pass the `-l` option to change which host
`pult-server` binds to:

```sh
~ $ pult-server -l 127.1.1.1
```

For more details on the server implementation, see the [server
documentation](http://cmcenroe.me/pult/server.html).

## Similar projects

* [Pow](http://pow.cx/)
  * Advantages:
    * Runs as regular user
    * Allows serving the same application from multiple domains (not
      just subdomains)
  * Disadvantages:
    * OS X only
    * Rack-specific
    * Requires symlinking (explicit setup)
    * Takes over control of your server processes

## Compatibility

Pult is compatible with Linux systems using `/etc/resolv.conf` and with
Mac OS X using `/etc/resolver`.

## License

Copyright © 2014, Curtis McEnroe <programble@gmail.com>

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted, provided that the above
copyright notice and this permission notice appear in all copies.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.


ChangeLog.md
# Change Log

## 2.1.1 (2015-01-02)

 * Respond to `A` record questions for unassigned domains with
   `NXDOMAIN`

## 2.1.0 (2014-11-01)

 * Add `.pultrc` support
 * Deprecate `.pult`

## 2.0.3 (2014-10-18)

 * Send `X-Forwarded` headers when proxying

## 2.0.2 (2014-05-19)

 * Use simpler `process.platform` instead of `os.platform()`
 * Use `readStream.resume()` to let the stream end without reading data

## 2.0.1 (2014-05-18)

 * Fix reading `.pult`

## 2.0.0 (2014-05-18)

 * Add Docco documentation
 * Add client library
 * Automatically start server from command line
 * Add built-in static file server
 * Add `--help` to `pult-server`
 * Add long options to `pult-server`
 * Send `Content-Length` header for responses from `pult-server`
 * Support `HEAD` requests to `pult-server`
 * Load `status.html` once
 * Use Bootstrap 3.1.1 on `pult-server` status page

## 1.0.1 (2014-05-11)

 * Update dependencies
 * Remove `console.log` from `pult-server`

## 1.0.0 (2014-05-11)

 * Log domain and port before starting command
 * Prefix `pult` output with `[pult]`
 * Add `pult` `-q` / `--quiet` option
 * Add `pult` long options and help text

## 0.7.0 (2014-02-04)

* Add HTML status page on http://pult.dev/
* Blacklist `favicon.ico` from becoming a domain

## 0.6.0 (2014-02-03)

* Warn if `pult` and `pult-server` are different versions
* Exit `pult-server` even if nameserver line not in `/etc/resolv.conf`

## 0.5.1 (2014-02-02)

* Respond with `NOTIMP` to unsupported DNS queries

## 0.5.0 (2014-02-01)

* Remove IPv6 listening from `pult-server`
* List next port to assign in `GET /` response
* Add `pult-server` `-p` and `-l` options

## 0.4.1 (2014-01-29)

* Respond with `A` and `AAAA` if DNS question type is `ANY`

## 0.4.0 (2014-01-28)

* Add `-p` and `-P` options
* Set first port to 7001

## 0.3.2 (2014-01-27)

* Ignore port on Host header (fixes bug if server is accessed at
  `http://foo.dev:80`)

## 0.3.1 (2014-01-27)

* Always proxy to IPv4 (fixes bug where POST body was not being sent to
  IPv4 after it was already sent to IPv6)

## 0.3.0 (2014-01-27)

* Fork `pult-server` to background
* Kill `pult-server` with `pult -k`

## 0.2.2 (2014-01-27)

* Exit after removing `/etc/resolver/dev`

## 0.2.1 (2014-01-27)

* Add OS X support (using `/etc/resolver/dev`)
* Return 502 for unknown hosts

## 0.2.0 (2014-01-27)

* Return 502 with JSON on proxy error
* Add IPv6 HTTP server
* Add websocket support
* Add support for subdomains

## 0.1.0 (2014-01-26)

Initial release


pult.js
#!/usr/bin/env node

var fs = require('fs');
var http = require('http');
var spawn = require('child_process').spawn;

function log() {
  if (!log.quiet)
    console.error('[pult]', Array.prototype.join.call(arguments, ' '));
}

var argv = process.argv.slice(2);

var name = '', method = 'PORT';

while (argv[0] && argv[0][0] == '-') {
  switch (argv[0]) {
  case '-k':
  case '--kill':
    return http.request({ hostname: 'pult.dev', method: 'DELETE' },
      function httpDelete(res) {
        process.exit();
      }).end();
  case '-n':
  case '--name':
    name = argv[1];
    argv = argv.slice(2);
    break;
  case '-p':
  case '--port':
  case '-P':
  case '--Port':
    method = argv.shift();
    break;
  case '-q':
  case '--quiet':
    log.quiet = argv.shift();
    break;
  case '-h':
  case '--help':
    log.quiet = false;
    log('-k, --kill              Kill pult-server');
    log('-n, --name "name"       Set domain name');
    log('-p, --port              Set port by passing -p to command');
    log('-P, --Port              Set port by passing -P to command');
    log('-q, --quiet             Disable logging');
    return;
  default:
    log('unknown option', argv[0]);
    process.exit(1);
  }
}

if (!name && argv.length > 0) {
  fs.readFile('.pult', { encoding: 'utf8' }, function readDotPult(err, data) {
    if (!err)
      name = data.trim();
    else
      name = process.cwd().split('/').reverse()[0];
    getPort();
  });
} else {
  getPort();
}

function getPort() {
  http.get('http://pult.dev/' + name, function httpGet(res) {
    var package = require('./package.json');
    if (res.headers['x-pult-version'] != package.version)
      log('warning: pult-server has different version:',
          res.headers['x-pult-version'], '!=', package.version);

    res.setEncoding('utf8');
    res.on('data', function httpGetData(data) {
      var ports = JSON.parse(data);
      for (var host in ports)
        if (name)
          spawnWithPort(ports[host]);
        else
          log(host, ports[host]);
    });
  });
}

function spawnWithPort(port) {
  if (method == 'PORT')
    process.env.PORT = port;
  else
    argv.push(method, port);
  log('http://' + name + '.dev', '-->', 'http://localhost:' + port);
  spawn(argv[0], argv.slice(1), { stdio: 'inherit' });
}


example.js
// # Pult
//
// ## Client Library Example
//
// This is an example of a Node HTTP server using Pult to bind to a local
// `.dev` domain.
//

var http = require('http');
var pult = require('pult');

// The server will be available at `http://hello.dev`.
var SERVER_NAME = 'hello';
// The port to fall back to if Pult is not available.
var DEFAULT_PORT = 8080;

// Create a simple HTTP server that responds with "Hello, World!" to every
// request.
var server = http.createServer(function(req, res) {
  var body = 'Hello, World!';
  res.writeHead(200, {
    'Content-Type': 'text/plain',
    'Content-Length': body.length
  });
  res.write(body);
  res.end();
});

// Only try to use Pult in development.
if (process.env.NODE_ENV == 'development') {
  // Check if Pult is running.
  pult.serverStatus(function(err, up) {
    if (err) throw err;
    // If the Pult server is up, retrieve a port number for `SERVER_NAME`
    // from it.
    if (up) {
      pult.getPort(SERVER_NAME, function(err, port, domain) {
        if (err) throw err;
        // Bind the HTTP server to the assigned port.
        server.listen(port);
        // Output a link to the `.dev` domain.
        console.log('http://' + domain);
      });
    // If the Pult server is not running, bind the HTTP server to the `PORT`
    // environment variable or the default development port.
    } else {
      var port = process.env.PORT || DEFAULT_PORT;
      server.listen(port);
      // Output a link to the server.
      console.log('http://localhost:' + port);
    }
  });
// In production, bind the HTTP server to the `PORT` environment variable.
} else {
  server.listen(process.env.PORT);
}

// ### Starting the Pult server automatically
//
// Though not recommended because some developers may not want to run the Pult
// server, it is possible to automatically start it before your server starts.
//

// Callback to start the server once Pult is running.
function startServer() {
  pult.getPort(SERVER_NAME, function(err, port, domain) {
    if (err) throw err;
    server.listen(port);
    console.log('http://' + domain);
  });
}

// Check if the Pult server is already running.
pult.serverStatus(function(err, up) {
  if (err) throw err;
  // If Pult is already running, start the server.
  if (up) {
    startServer();
  // Otherwise, spawn the Pult server.
  } else {
    pult.spawnServer(function(err) {
      if (err) throw err;
      // Then start the HTTP server.
      startServer();
    });
  }
});


lib/index.js
// # Pult
//
// Access local servers on `.dev` domains.
//
// [GitHub](https://github.com/programble/pult)
//
// ## Contents
//
//  - [Client Library](index.html)
//  - [Client Library Example](example.html)
//  - [Command Line](pult.html)
//  - [Server](server.html)
//  - [Static File Server](static.html)
//
// ## Client Library
//
// ```
// npm install --save-dev pult
// ```
//
// ```
// var pult = require('pult');
// ```
//
// The Pult client library allows you to start up your Node server on a local
// `.dev` domain. For an example Node server using Pult, see the [Client
// Library Example](example.html).
//

var path  = require('path');
var spawn = require('child_process').spawn;
var dns   = require('dns');
var http  = require('http');

var package = require('../package.json');

exports = module.exports = {
  // ### Server Status
  //
  // Get the running status of `pult-server`.
  //
  // Callback should take `(err, up)`, where `up` is a boolean indicating if
  // `pult-server` is available on `pult.dev`.
  //
  serverStatus: function(done) {
    dns.lookup('pult.dev', function(err, address, family) {
      if (err && err.code == 'ENOTFOUND') done(null, false);
      else done(err, true);
    });
  },

  // ### Spawn Server
  //
  // Spawn `pult-server`.
  //
  // `args` is an optional array of arguments to be passed to `pult-server`.
  //
  // Callback should take `(err)` and is called once `pult-server` is available
  // on `pult.dev`.
  //
  spawnServer: function(args, done) {
    if (typeof args == 'function') {
      done = args;
      args = [];
    }

    args.unshift(path.join(__dirname, '..', 'bin', 'server.js'));
    var serverProcess = spawn(process.execPath, args, {
      stdio: 'inherit'
    });

    serverProcess.on('exit', function(code, signal) {
      if (code != 0) {
        var err = new Error('pult-server failed to start');
        err.exitCode = code;
        return done(err);
      }

      var tries = 0;
      var serverWait = function(err, up) {
        if (err) return done(err);
        if (up) return done();
        if (++tries == 50) {
          var err = new Error(
            'pult-server started but was unavailable after 5 seconds'
          );
          return done(err);
        }
        setTimeout(function() {
          exports.serverStatus(serverWait)
        }, 100);
      };
      serverWait();
    });
  },

  // ### Kill Server
  //
  // Kill `pult-server` by sending HTTP DELETE / to `http://pult.dev`.
  //
  // Callback should take `(err)` and is called when the HTTP response is
  // received.
  //
  killServer: function(done) {
    http.request({
      hostname: 'pult.dev',
      method: 'DELETE'
    }, function(res) {
      done();
    }).on('error', done).end();
  },

  // ### Server Version
  //
  // Get the version of the running `pult-server` instance.
  //
  // Callback should take `(err, serverVersion, clientVersion)`.
  //
  serverVersion: function(done) {
    http.get('http://pult.dev', function(res) {
      done(null, res.headers['x-pult-version'], package.version);
    }).on('error', done);
  },

  // ### Get
  //
  // Send an HTTP GET request to `pult.dev`.
  //
  // This function is not recommended for public use, instead you should use
  // `getPorts` or `getPort`, which call this function.
  //
  // `path` is the path to GET.
  //
  // Callback should take `(err, json)`, where `json` is the JSON response from
  // `pult-server`.
  //
  get: function(path, done) {
    http.get('http://pult.dev' + path, function(res) {
      res.setEncoding('utf8');
      res.on('data', function(data) {
        done(null, JSON.parse(data));
      });
    }).on('error', done);
  },

  // ### Get Ports
  //
  // Retrieve domain names and corresponding ports from `pult-server`.
  //
  // Callback should take `(err, ports)`, where `ports` is an object containing
  // domain name keys and port number values. Note that `pult.dev` as well as
  // `next` (indicating the next port number to assign) are included in this
  // object.
  //
  getPorts: function(done) {
    exports.get('/', done);
  },

  // ### Get Port
  //
  // Retrieve the port number assigned to a domain from `pult-server`.
  //
  // `name` is the domain name, without `.dev`. For example, to retrieve the
  // port number assigned to `foobar.dev`, call `getPort` with `'foobar'`.
  //
  // Callback should take `(err, port, domain)`, where `port` is the assigned
  // port number and `domain` is the full domain name (with `.dev`).
  //
  getPort: function(name, done) {
    exports.get('/' + name, function(err, ports) {
      for (var domain in ports) {
        return done(null, ports[domain], domain);
      }
      return done(new Error('did not receive port'));
    });
  }
};


lib/static.js
// # Pult
//
// ## Static File Server
//
// A simple static file HTTP server that the [Pult command line](pult.html)
// runs for the `--static` option. Used to easily make static assets available
// on a local `.dev` domain.
//
// This could be extended in future versions to serve directory listings as
// well as files.
//

var http = require('http');
var path = require('path');
var mime = require('mime');
var fs   = require('fs');

// ### Log
//
// Log output to stdout with a prefix.
//
function log() {
  console.log('[static]', Array.prototype.join.call(arguments, ' '));
}

// ### Write Status
//
// Write an HTTP status code and its status text to the body of an HTTP
// response.
//
function writeStatus(status, headers) {
  if (!headers) headers = {};
  this.logType = headers['Content-Type'] = 'text/plain';
  this.logLength = headers['Content-Length'] = http.STATUS_CODES[status].length;
  this.writeHead(status, headers);
  if (this.req.method != 'HEAD')
    this.write(http.STATUS_CODES[status]);
}

// ### End
//
// End the HTTP response (by calling the original `res.end()` and log the
// request.
//
function end() {
  this._end();
  log(
    new Date().toISOString(),
    this.req.method,
    this.req.url,
    this.statusCode,
    http.STATUS_CODES[this.statusCode],
    this.logType,
    this.logLength
  );
}

// ### HTTP Server
//
var server = http.createServer(function(req, res) {
  // Make request object accessible to response methods.
  res.req = req;
  // Assign extra methods to response object.
  res.writeStatus = writeStatus;
  res._end = res.end;
  res.end = end;

  // Only GET and HEAD are supported.
  if (req.method != 'GET' && req.method != 'HEAD') {
    res.writeStatus(405, { 'Allow': 'GET, HEAD' });
    return res.end();
  }

  // Serve files from the working directory.
  var filePath = path.resolve(path.join(process.cwd(), req.url));

  // Do not serve paths outside of the working directory (i.e. `..`).
  if (filePath.indexOf(process.cwd()) != 0) {
    res.writeStatus(403);
    return res.end();
  }

  // Attempt to open the file for reading.
  var readStream = fs.createReadStream(filePath);

  readStream.on('error', function(err) {
    // If the path points to a directory, redirect to `index.html` in that
    // directory.
    if (err.code == 'EISDIR') {
      res.writeStatus(302, { 'Location': path.join(req.url, 'index.html') });
      return res.end();
    }

    // Return 404 if the path does not exist.
    if (err.code == 'ENOENT') {
      res.writeStatus(404);
      return res.end();
    }

    // Return 403 if the path cannot be accessed.
    if (err.code == 'EACCES') {
      res.writeStatus(403);
      return res.end();
    }

    // If any other error occurred, return 500 and log the error.
    res.writeStatus(500);
    log(err.stack);
    return res.end();
  });

  readStream.on('open', function(fd) {
    // Stat the file to get its size and modification date.
    fs.fstat(fd, function(err, stats) {
      if (err) {
        res.writeStatus(500);
        log(err.stack);
        return res.end();
      }

      // Look up the MIME type of the path.
      var mimeType = mime.lookup(filePath);

      // Set response headers for type, length, and modification time.
      var headers = {
        'Content-Type': mimeType,
        'Content-Length': stats.size,
        'Last-Modified': stats.mtime.toUTCString()
      };

      // Save type and length for request logging.
      res.logType = mimeType;
      res.logLength = stats.size;

      // If the file has not been modified since the value of the
      // `If-Modified-Since` request header, return 304.
      if (req.headers['if-modified-since']) {
        var since = new Date(req.headers['if-modified-since']);
        if (stats.mtime <= since) {
          // Let data flow but don't read it.
          readStream.resume();
          res.writeHead(304, headers);
          return res.end();
        }
      }

      // Send the headers for the requested file.
      res.writeHead(200, headers);

      // Write the file's contents to the response body.
      if (req.method != 'HEAD') {
        readStream.pipe(res);
      } else {
        // Let data flow but don't read it.
        readStream.resume();
        res.end();
      }
    });
  });
});

module.exports = server;


pult-server.js
#!/usr/bin/env node

var spawn = require('child_process').spawn;
var os = require('os');
var fs = require('fs');
var dns = require('native-dns');
var http = require('http');
var httpProxy = require('http-proxy');

if (process.getuid() != 0) {
  console.log('spawning with sudo...');

  var args = process.argv.slice(1);
  args.unshift(process.execPath);
  spawn('sudo', args, { stdio: 'inherit' });
  return;
}

if (process.argv.indexOf('-f') == -1) {
  console.log('forking to background...');

  var args = process.argv.slice(1);
  args.splice(1, 0, '-f');
  spawn(process.execPath, args, { stdio: 'ignore', detached: true }).unref();
  return;
}

var listenHost = '127.0.0.1';
var firstPort = 7001;

var argv = process.argv.slice(2);
while (argv[0]) {
  var arg = argv.shift();
  switch (arg) {
  case '-p':
    firstPort = argv.shift();
    break;
  case '-l':
    listenHost = argv.shift();
    break;
  case '-f':
    break;
  default:
    console.log('unknown option ' + arg);
    process.exit(1);
  }
}

var ports = {
  'pult.dev': 80,
  next: firstPort
};

function getPort(host) {
  while (!ports[host] && host.indexOf('.') != -1)
    host = host.slice(host.indexOf('.') + 1);
  return ports[host];
}

var dnsServer = dns.createServer();

dnsServer.on('request', function dnsRequest(req, res) {
  var name = req.question[0].name;
  var type = dns.consts.QTYPE_TO_NAME[req.question[0].type];
  if (getPort(name) && (type == 'A' || type == 'ANY')) {
    res.answer.push(dns.A({
      name: name,
      address: listenHost,
      ttl: 600
    }));
  } else {
    res.header.rcode = dns.consts.NAME_TO_RCODE.NOTIMP;
  }
  res.send();
});

dnsServer.serve(53, listenHost);

var resolvConfLine = 'nameserver ' + listenHost;

if (os.platform() == 'darwin') {
  fs.mkdir('/etc/resolver', function mkdirResolver(err) {
    console.log('adding /etc/resolver/dev');
    fs.writeFile('/etc/resolver/dev', resolvConfLine,
      function writeResolver(err) { if (err) throw err; });
  });
} else {
  fs.readFile('/etc/resolv.conf', { encoding: 'utf8' },
    function readResolvConf(err, data) {
      if (err) throw err;
      if (data.indexOf(resolvConfLine) == -1) {
        console.log('adding ' + resolvConfLine + ' to /etc/resolv.conf');
        fs.writeFile('/etc/resolv.conf', resolvConfLine + '\n' + data,
          function writeResolvConf(err) { if (err) throw err; });
      }
  });
}

process.on('SIGINT', onExit);
process.on('SIGTERM', onExit);
function onExit() {
  if (os.platform() == 'darwin') {
    console.log('removing /etc/resolver/dev');
    fs.unlink('/etc/resolver/dev', function(err) {});
    process.exit();
  } else {
    fs.readFile('/etc/resolv.conf', { encoding: 'utf8' }, function(err, data) {
      if (err) throw err;
      if (data.indexOf(resolvConfLine) == 0) {
        console.log('removing ' + resolvConfLine + ' from /etc/resolv.conf');
        fs.writeFile('/etc/resolv.conf', data.replace(resolvConfLine + '\n', ''),
          function(err) {
            if (err) throw err;
            process.exit();
          });
      } else {
        process.exit();
      }
    });
  }
}

var proxy = httpProxy.createProxyServer({});

function resJSON(res, code, obj) {
  var package = require('./package.json');
  res.writeHead(code, {
    'X-Pult-Version': package.version,
    'Content-Type': 'application/json'
  });
  res.write(JSON.stringify(obj));
  res.end();
}

function httpRequest(req, res) {
  var host = req.headers.host;
  host = host.split(':')[0]; // Ignore port
  var port = getPort(host);

  if (host == 'pult.dev') {
    if (req.method == 'DELETE' && req.url == '/') {
      resJSON(res, 200, ports);
      return onExit();
    }
    if (req.method != 'GET')
      return resJSON(res, 400, { method: req.method });

    if (req.url == '/favicon.ico')
      return resJSON(res, 404, { url: req.url });

    var name = req.url.slice(1);

    if (name) {
      name += '.dev';
      port = {};
      port[name] = ports[name] || (ports[name] = ports.next++);
      resJSON(res, 200, port);
    } else {
      if (req.headers.accept && req.headers.accept.indexOf('text/html') == 0) {
        fs.readFile(__dirname + '/status.html', { encoding: 'utf8' },
          function readHtml(err, data) {
            if (err) return resJSON(res, 500, err);
            res.writeHead(200, { 'Content-Type': 'text/html' });
            res.write(data);
            res.end();
          });
      } else {
        resJSON(res, 200, ports);
      }
    }
  } else if (port) {
    proxy.web(req, res, { target: 'http://127.0.0.1:' + port },
      function proxyWebError4(err) {
        resJSON(res, 502, err);
      });
  } else {
    resJSON(res, 502, { host: host });
  }
}

function httpUpgrade(req, socket, head) {
  var port = getPort(req.headers.host);
  if (port) {
    proxy.ws(req, socket, head, { target: 'ws://127.0.0.1:' + port },
      function proxyWsError4(err) {
        resJSON(res, 502, err);
      });
  } else {
    resJSON(res, 502, { host: req.headers.host });
  }
}

var httpServer = http.createServer();
httpServer.on('request', httpRequest);
httpServer.on('upgrade', httpUpgrade);
httpServer.listen(80, listenHost);
